// @antv/l7-core/es/services/renderer/passes/ClearPass.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _get from "@babel/runtime/helpers/get";
import _inherits from "@babel/runtime/helpers/inherits";

var _dec, _class;

import { injectable } from 'inversify';
import BaseNormalPass from './BaseNormalPass';
var ClearPass = (_dec = injectable(), _dec(_class = function (_BaseNormalPass) {
  _inherits(ClearPass, _BaseNormalPass);

  function ClearPass() {
    _classCallCheck(this, ClearPass);

    return _possibleConstructorReturn(this, _getPrototypeOf(ClearPass).apply(this, arguments));
  }

  _createClass(ClearPass, [{
    key: "getName",
    value: function getName() {
      return 'clear';
    }
  }, {
    key: "init",
    value: function init(layer, config) {
      _get(_getPrototypeOf(ClearPass.prototype), "init", this).call(this, layer, config);
    }
  }, {
    key: "render",
    value: function render() {
      this.rendererService.clear({
        color: [0, 0, 0, 0],
        depth: 1,
        framebuffer: null
      });
    }
  }]);

  return ClearPass;
}(BaseNormalPass)) || _class);
export { ClearPass as default };
