// @antv/l7-layers/es/raster/index.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/inherits";
import BaseLayer from '../core/BaseLayer';
import RasterModels from './models/index';

var RaterLayer = function (_BaseLayer) {
  _inherits(RaterLayer, _BaseLayer);

  function RaterLayer() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, RaterLayer);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(RaterLayer)).call.apply(_getPrototypeOf2, [this].concat(args)));
    _this.type = 'RasterLayer';
    return _this;
  }

  _createClass(RaterLayer, [{
    key: "buildModels",
    value: function buildModels() {
      var modelType = this.getModelType();
      this.layerModel = new RasterModels[modelType](this);
      this.models = this.layerModel.buildModels();
    }
  }, {
    key: "getConfigSchema",
    value: function getConfigSchema() {
      return {
        properties: {
          opacity: {
            type: 'number',
            minimum: 0,
            maximum: 1
          }
        }
      };
    }
  }, {
    key: "getDefaultConfig",
    value: function getDefaultConfig() {
      var type = this.getModelType();
      var defaultConfig = {
        raster: {},
        raster3d: {}
      };
      return defaultConfig[type];
    }
  }, {
    key: "getModelType",
    value: function getModelType() {
      return 'raster';
    }
  }]);

  return RaterLayer;
}(BaseLayer);

export { RaterLayer as default };
